src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelInternal.java

```diff
@@ -246,24 +201,28 @@ public String getFullyQualifiedTableName() {
    * @return a ChannelData object
    */
   ChannelData<T> getData() {
-    return this.rowBuffer.flush();
+    ChannelData<T> data = this.rowBuffer.flush();
+    if (data != null) {
+      data.setChannelContext(channelContext);
+    }
+    return data;
   }
 
   /** @return a boolean to indicate whether the channel is valid or not */
   @Override
   public boolean isValid() {
-    return this.isValid;
+    return this.rowBuffer.isValid();
   }
 
   /** Mark the channel as invalid, and release resources */
   void invalidate(String message) {
-    this.isValid = false;
+    this.rowBuffer.invalidate();
```

Comment:
Similar concern here, the isValid should be on a channel instead of a rowBuffer

Reply:
This was one of the key points of refactoring.
The `isValid` flag is used in the `ArrowRowBuffer#close` method.
As I understand there is 1-1 relationship between channel and its buffer.
If channel is invalid then the buffer as well.
I would not touch it if the `isValid` flag was not used in `ArrowRowBuffer#close`.

Another option is that `ArrowRowBuffer#close` takes `isValid` as argument
but it is a volatile field and I was not sure about all consequences of capturing its value in a local function argument
before calling `ArrowRowBuffer#close`. If you say it is safe, it can be done like this.

Reply:
now part of the `ChannelState`